### ISSUE TITLE
Stuck with Connecting state

### ISSUE THREAD

**Issue by kanat** [Polidea/RxAndroidBle#178]
[conn_stuck_key.txt](https://github.com/Polidea/RxAndroidBle/files/947274/conn_stuck_key.txt)
### Summary
Stuck with Connecting state
#### Preconditions
I'm just calling `device.establishConnection(false)`
#### Steps to reproduce actual result
1. `device.establishConnection(false)`
2. Unsubcribe
3. `device.establishConnection(false)`
4. Repeat again from 1 N times
#### Actual result
CONNECTING state

`04-21 16:48:15.335 25090-25183/ru.starline.key E/ConnectionManager: (Key-1-1:2184) [connect] RxBleDeviceImpl{bluetoothDevice=StarLine BLE(F8:5B:60:89:CD:8A)} (autoConnect='false')`

And I get

`04-21 16:48:50.354 25090-25219/ru.starline.key E/ConnectionManager: (RxComputationScheduler-3:2194) [connect] error: BleGattCallbackTimeoutException{macAddress=null, bleGattOperationType=BleGattOperation{description='CONNECTION_STATE'}}`

I don't understand why mac address is null

Also you can see at logs that Connection was established:
`04-21 16:48:15.357 25090-25112/ru.starline.key D/BluetoothGatt: onClientConnectionState() - status=0 clientIf=6 device=F8:5B:60:89:CD:8A`
`04-21 16:48:15.357 25090-25112/ru.starline.key D/RxBle#BluetoothGatt: onConnectionStateChange newState=2 status=0`

#### Expected result

CONNECTED state
<br />

**Comment by dariuszseweryn** [comment]
Nice one. It seems to be a race condition in the `RxBleRadioOperationConnect`. I am writing a fix.

**Comment by kanat** [comment]
Is it possible to configure timeout value for ble operation?
Now it takes about 35 seconds to get BleGattCallbackTimeoutException.

**Comment by dariuszseweryn** [comment]
From connection observable you can unsubscribe at any moment and it should be handled by the library.

**Comment by kanat** [comment]
Unsubscribing is not helping. 

The same result.

**Comment by kanat** [comment]
As Understand the problem occurs here in RxBleRadioOperationDisconnect (line 62):

```
if (bluetoothGatt == null) {
    RxBleLog.w("Disconnect operation has been executed but GATT instance was null.");
    onCompleted();
}
```

After calling `onCompleted` the method `releaseRadio` is never called.

**Comment by dariuszseweryn** [comment]
True. Another thing that did not manifest before.

**Comment by timmiller17** [comment]
I think I am struggling with this as well.  How is the fix coming along?

**Comment by dariuszseweryn** [comment]
Fix for the original cause is ready and waiting on a private PR to be reviewed by @uKL but he has a quite full plate at the moment. I hope that he will have a moment tomorrow to check it out. The fix for the manifestation (Disconnect) I will make tomorrow.

**Comment by kanat** [comment]
Is there any way to push my own **ConnectionModule** into the RxAndroidBle Graph?
(Just to provide fixed versions of operations, until official release will be published)

**Comment by dariuszseweryn** [comment]
Unfortunately - There is no possibility to do this. You would have to compile your own version from the sources.

**Comment by kanat** [comment]
What a pity.
That would be nice to have)

**Comment by kanat** [comment]
> It seems to be a race condition in the RxBleRadioOperationConnect

@dariuszseweryn could you tell what was the problem at RxBleRadioOperationConnect?

**Comment by dariuszseweryn** [comment]
`RxBleGattCallback` is emitting `onConnectionStateChange()` before the subscription chain is fully established.

**Comment by dariuszseweryn** [comment]
@kanat I have pushed fixes for all issues found by you. These should be available in the `1.3.0-SNAPSHOT` artifact shortly. Could you give it a try?

**Comment by kanat** [comment]
@dariuszseweryn yes, sure, I can try it with real device. 
Where/How can I get `1.3.0-SNAPSHOT` artifact?

**Comment by dariuszseweryn** [comment]
See https://github.com/Polidea/RxAndroidBle
`Snapshot` paragraph

**Comment by kanat** [comment]
Actually I've tried it with my own fixes, seems it works :)
Anyway I'll try `1.3.0-SNAPSHOT` as well.

**Comment by kanat** [comment]
@dariuszseweryn It works, thanks.

**Comment by kanat** [comment]
@dariuszseweryn, is it possible to get frozen version of `1.3.0-SNAPSHOT`, to avoid situation where my own app can be affected by new changes which can arrive with a modified SNAPSHOT?
Or maybe it's better to compile RxAndroidBle by my own, until the new release will be published?

**Comment by dariuszseweryn** [comment]
Great news. I will try to publish `1.2.1` tomorrow. You are welcome.

Your own version will be more bulletproof. SNAPSHOTs tend to change but regression is rarely an issue.

**Comment by dariuszseweryn** [comment]
@kanat I have just released `1.2.1`
It should be available shortly.

**Comment by kanat** [comment]
@dariuszseweryn Great news, thanks!
